File: src/main/java/com/agilesparks/rubikscube/solver/Solver.java

package com.agilesparks.rubikscube.solver;

import com.agilesparks.rubikscube.cube.Cube;
import com.agilesparks.rubikscube.cube.NewCube;

public class Solver {

    public Solver() {
    }

    public Solution solve(Cube p_rubik, RotationTree p_firstTree, RotationTree p_secondTree, RotationTree p_thirdTree) {

        int l_numberOfCubicleInPlace;
        NewCube l_permutation = Cube.getPermutationFromCube(p_rubik);
        SolutionManager l_solutionManager = new SolutionManager();
        Solution l_solutionToDev;

        RotationLinkedList l_rotationLinkedList = new RotationLinkedList();

        int l_floor = getTargetFloor(l_permutation);
        l_numberOfCubicleInPlace = Cube.getValue(l_permutation, l_floor);

        l_solutionManager.addSolution(l_rotationLinkedList, l_permutation, null, l_numberOfCubicleInPlace,l_floor);
        while ((l_solutionToDev = l_solutionManager.getBestUndeveloped()) != null &&
                l_solutionManager.getBestValue() < 40) {
            int targetFloor = getTargetFloor(l_solutionToDev.getPermutation());
           System.out.format("Searching %d",Cube.getValue(l_solutionToDev.getPermutation(), targetFloor));
            if (l_solutionManager.getBestValue()>Cube.getValue(l_solutionToDev.getPermutation(), targetFloor)+14)
            {
                System.out.println("Couldn't Find a Solution");
                return l_solutionManager.getBest();
            }
            if (targetFloor == 1)
                findBetterSolution(l_solutionToDev, p_firstTree, l_solutionManager,targetFloor );
            if (targetFloor == 2)
                findBetterSolution(l_solutionToDev, p_secondTree, l_solutionManager,targetFloor );
            if (targetFloor == 3)
                findBetterSolution(l_solutionToDev, p_thirdTree, l_solutionManager,targetFloor );

            l_floor = getTargetFloor(l_solutionManager.getBestValue());

            System.out.format("Floor=%d, Best yet:%d\n", l_floor, l_solutionManager.getBestValue());
            // l_solutionManager.getBest().print();


        }

        return l_solutionManager.getBest();

    }

    public int getTargetFloor(NewCube p_permutation) {
        if (Cube.getValue(p_permutation, 1) >= 16) {
            if (Cube.getValue(p_permutation, 2) < 24) {
                return 2;
            } else {
                return 3;
            }
        } else {
            return 1;
        }
    }

    public int getTargetFloor(int p_value) {
        if (p_value >= 16) {
            if (p_value < 24) {
                return 2;
            } else {
                return 3;
            }
        } else {
            return 1;
        }
    }




    private void findBetterSolution(Solution p_solution, RotationTree p_tree, SolutionManager p_solutionManager,
                                    int p_floor) {
        Cube l_rubik = new Cube();
        NewCube l_permutation = p_solution.getPermutation().getCopy();
        RotationLinkedList l_rotationLinkedList = new RotationLinkedList();
        int l_minimumValue = Cube.getValue(l_permutation, p_floor);

//	if (l_minimumValue < 8)
//		l_minimumValue = 8;
        l_rubik.reset(l_permutation);
        searchTree(l_minimumValue-4, p_tree, l_rubik, p_solutionManager,
                p_solution, p_floor, 0);
    }


    public void searchTree(int p_minimumValue, RotationTree p_tree,
                           Cube p_rubik,SolutionManager p_solutionManager,
                           Solution p_prevSolution, int p_floor, int depth) {
        if (p_minimumValue<2) p_minimumValue = 2;
        NewCube l_permutation = Cube.getPermutationFromCube(p_rubik).getCopy();
        Cube l_rubik = new Cube();
        for (int i=0;i<p_tree.getSize();i++){
            RotationLinkedList l_rotationLinkedList = p_tree.getRotationLinkedList(i);
            if (l_rotationLinkedList != null) {
                l_rubik.reset(l_permutation);
                for (int j=0;j<l_rotationLinkedList.size();j++)
                    l_rubik.rotateFace(l_rotationLinkedList.get(j).getFace(), l_rotationLinkedList.get(j).getDirection() );
                NewCube l_resultPermutation = Cube.getPermutationFromCube(l_rubik).getCopy();

                if (Cube.getValue(l_resultPermutation, p_floor) >= p_minimumValue) {
                    p_solutionManager.addSolution(l_rotationLinkedList, l_resultPermutation, p_prevSolution, Cube.getValue(l_resultPermutation, p_floor), p_floor);
               //     if (depth == 1) System.out.format("Hi,value=%d\n",l_resultPermutation.getValue(p_floor));
                }
                if (p_floor==3 && depth==0) {
                  //  System.out.println("Hi");
                    searchTree(p_minimumValue, p_tree, l_rubik, p_solutionManager,
                            new Solution(l_rotationLinkedList, l_resultPermutation, p_prevSolution), p_floor, 1);

                }
            }
        }
    }



}

File: original_2001_cpp/Solver.cpp
/***************************************************************************
                          Solver.h  -  description
                             -------------------
    begin                : Fri May 25 2001
    copyright            : (C) 2001 by yaki koren
    email                :
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "Solver.h"


void Solver::solve(Rubik *p_rubik)
{
	int l_value,l_onlyFirstFloor = 0;
	Permutation *l_permutation = p_rubik->getPermutation();
	SolutionManager *l_solutionManager = new SolutionManager();
	Solution *l_solutionToDev,*l_finalSolution;
 	RotationTree *l_tree = loadSearchTree();
  Process *l_process = new Process();;
	l_value = l_permutation->getValue(0);
	if (l_value < 16)
	{
		l_onlyFirstFloor = 1;
		l_value = l_permutation->getValue(1);
	}
		
 	
 	l_solutionManager->addSolution(l_process,l_permutation,NULL,l_value);
	while ((l_solutionToDev = l_solutionManager->getBestUndeveloped())!=NULL &&
					l_solutionManager->getBestValue()<40)
	{
		develop(l_solutionToDev,l_tree,l_solutionManager,l_onlyFirstFloor);
		printf("Best yet:%d\n",l_solutionManager->getBestValue());	
		if (l_solutionManager->getBestValue()>=14)
			l_onlyFirstFloor = 0;
  }
  l_finalSolution = l_solutionManager->getBest();
  printf("the Solution found values is:%d\n",l_finalSolution->getPermutation()->getValue(l_onlyFirstFloor));
  printf("the process is:\n");
  l_finalSolution->print();
  l_finalSolution->applyToRubik(p_rubik);
  getchar();
  p_rubik->printCube();
}

void Solver::develop(Solution *p_solution,RotationTree *p_tree,SolutionManager *p_solutionManager,
											int p_onlyFirstFloor)
{
	Rubik *l_rubik = new Rubik();
	Permutation *l_permutation = p_solution->getPermutation()->getCopy();
	Process l_process;
	int l_minimumValue = l_permutation->getValue(p_onlyFirstFloor);
//	if (l_minimumValue < 8)
//		l_minimumValue = 8;
	l_rubik->set(l_permutation);
	searchTree(l_minimumValue,p_tree,l_rubik,&l_process,p_solutionManager,
								p_solution,p_onlyFirstFloor);
}


void Solver::searchTree(int p_minimumValue, RotationTree *p_tree,
								Rubik *p_rubik,Process *p_process,SolutionManager *p_solutionManager,
								Solution *p_prevSolution,int p_onlyFirstFloor)
{
	int i;
	Rotation l_rotation;
	RotationTree *l_nextSubTree;
	for (i=0;i<12;i++)
		if (p_tree->getRotation(i,&l_rotation))
		{
			if (!(p_process->isRedundant(l_rotation)))
			{
				p_process->addRotation(l_rotation);
				p_rubik->rotateFace(l_rotation);
				Permutation *l_currPermutation = p_rubik->getPermutation();
				if (l_currPermutation->getValue(p_onlyFirstFloor)>= p_minimumValue)
//					printf("found %d\n", l_currPermutation->getValue(p_onlyFirstFloor));
					p_solutionManager->addSolution(p_process, l_currPermutation,p_prevSolution,l_currPermutation->getValue(p_onlyFirstFloor));
				if ((l_nextSubTree = p_tree->getNext(i))!=NULL)
					searchTree(p_minimumValue,l_nextSubTree,p_rubik,p_process,
											p_solutionManager,p_prevSolution,p_onlyFirstFloor);
				delete l_currPermutation;
				p_process->removeRotation();
				p_rubik->rotateFace(l_rotation.getReverse());
	    }
		}
}

RotationTree *Solver::loadSearchTree()
{
 	Process *l_process = new Process();
 	RotationTree *l_tree = new RotationTree();
 	FILE *l_file;
 	l_file= fopen("permRubik.txt","r");
  loadRotationTreeFromFile(l_file, l_tree);
  loadRotationTreeFromStandart(l_tree,l_process,5);
  fclose(l_file);
  return l_tree;
}

void Solver::loadRotationTreeFromFile(FILE *p_file, RotationTree *p_tree)
{
 	Process *l_process = new Process();
 	while (l_process->reset(p_file))
 	{
 		p_tree->addProcess(*l_process);
 	}

}


void Solver::loadRotationTreeFromStandart(RotationTree *p_tree, Process *p_process,int p_depth)
{
	int i;
	if (p_depth == 0) return;
	for (i=0;i<12;i++)
		{
		if (p_process->isRedundant(Rotation(i)))
			continue;		
		p_process->addRotation(Rotation(i));
		p_tree->addProcess(*p_process);
		loadRotationTreeFromStandart(p_tree,p_process,p_depth-1);
    p_process->removeRotation();
  	}
  return;
}
